### ISSUE TITLE
Typescript doesn't catch missing fields on mapped optional types

### ISSUE THREAD

**Issue by mikeecb** [Microsoft/TypeScript#29632]
**TypeScript Version:**  3.3.0

**Code**

```typescript
interface A {
  numberContainer?: NumberContainer;
  arrayContainer?: ArrayContainer;
}   

interface NumberContainer {           
  type: 'number';
  a: number;                                           
  b: number;       
}     
   
interface ArrayContainer {
  type: 'array';
  c: Array<any>;           
  d: Array<any>;
}                                                     

interface InferredContainerNoType<T> {
  y: {
    // This nested mapped type causes errors to not be caught by the compiler
    // (e.g. missing 'type' field)
    [key in keyof Pick<T, Exclude<keyof T, 'type'>>]: {
      value: T[key]
    };
  }
}

interface InferredANoType {
  x: {
    [key in keyof A]: InferredContainerNoType<A[key]>;
  };
}

// Incorrectly compiles without error (should error on missing 'a' and 'c' fields)
const inferredANoType_bad: InferredANoType = {
  x: {
    numberContainer: {
      y: {
        // Missing 'a' field
        b: { value: 2 },
      }
    },
    arrayContainer: {
      y: {
        // Missing 'c' field
        d: { value: [2] },
      },
    },
  },
};
```

**Expected behavior:**

I expect `inferredANoType_bad` to fail with a compilation error indicating that the `a` and `b` fields on `inferredANoType_bad.x.numberContainer.y` and  `inferredANoType_bad.x.arrayContainer.y` should exist respectively.

**Actual behavior:**

Compilation succeeds without failure.

Note: If I make `numberContainer` and `arrayContainer` required fields (i.e. not optional) on interface `A`, compilation fails with the expected errors.

**Playground Link:** [Link](https://www.typescriptlang.org/play/#src=interface%20A%20%7B%0D%0A%20%20numberContainer%3F%3A%20NumberContainer%3B%0D%0A%20%20arrayContainer%3F%3A%20ArrayContainer%3B%0D%0A%7D%20%20%20%0D%0A%0D%0Ainterface%20NumberContainer%20%7B%20%20%20%20%20%20%20%20%20%20%20%0D%0A%20%20type%3A%20'number'%3B%0D%0A%20%20a%3A%20number%3B%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%0D%0A%20%20b%3A%20number%3B%20%20%20%20%20%20%20%0D%0A%7D%20%20%20%20%20%0D%0A%20%20%20%0D%0Ainterface%20ArrayContainer%20%7B%0D%0A%20%20type%3A%20'array'%3B%0D%0A%20%20c%3A%20Array%3Cany%3E%3B%20%20%20%20%20%20%20%20%20%20%20%0D%0A%20%20d%3A%20Array%3Cany%3E%3B%0D%0A%7D%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%0D%0A%0D%0Ainterface%20InferredContainerNoType%3CT%3E%20%7B%0D%0A%20%20y%3A%20%7B%0D%0A%20%20%20%20%2F%2F%20This%20nested%20mapped%20type%20causes%20errors%20to%20not%20be%20caught%20by%20the%20compiler%0D%0A%20%20%20%20%2F%2F%20(e.g.%20missing%20'type'%20field)%0D%0A%20%20%20%20%5Bkey%20in%20keyof%20Pick%3CT%2C%20Exclude%3Ckeyof%20T%2C%20'type'%3E%3E%5D%3A%20%7B%0D%0A%20%20%20%20%20%20value%3A%20T%5Bkey%5D%0D%0A%20%20%20%20%7D%3B%0D%0A%20%20%7D%0D%0A%7D%0D%0A%0D%0Ainterface%20InferredANoType%20%7B%0D%0A%20%20x%3A%20%7B%0D%0A%20%20%20%20%5Bkey%20in%20keyof%20A%5D%3A%20InferredContainerNoType%3CA%5Bkey%5D%3E%3B%0D%0A%20%20%7D%3B%0D%0A%7D%0D%0A%0D%0A%2F%2F%20Incorrectly%20compiles%20without%20error%20(should%20error%20on%20missing%20'a'%20and%20'c'%20fields)%0D%0Aconst%20inferredANoType_bad%3A%20InferredANoType%20%3D%20%7B%0D%0A%20%20x%3A%20%7B%0D%0A%20%20%20%20numberContainer%3A%20%7B%0D%0A%20%20%20%20%20%20y%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20b%3A%20%7B%20value%3A%202%20%7D%2C%0D%0A%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%7D%2C%0D%0A%20%20%20%20arrayContainer%3A%20%7B%0D%0A%20%20%20%20%20%20y%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20d%3A%20%7B%20value%3A%20%5B2%5D%20%7D%2C%0D%0A%20%20%20%20%20%20%7D%2C%0D%0A%20%20%20%20%7D%2C%0D%0A%20%20%7D%2C%0D%0A%7D%3B)

**Comment by dragomirtitian** [comment]
Actually this seems to be due to an error in your code. Without strict null checks the compiler raises an error as you expect it. When you turn on strict null checks, the types for the members of `A` turn into `NumberContainer | undefined` and `ArrayContainer | undefined`. This means that `x` is typed as `InferredContainerNoType<NumberContainer | undefined>`. Since `keyof` a union type only yields common memebers and `undefined` has no members, `keyof NumberContainer | undefined` will be `never`.  This means the type of `y` will actually be `{}` and not the type you expect. 

This code will trigger the error you expect (it explicitly e excludes  `undefined` from `A[key]`):

```ts
    interface A {
      numberContainer?: NumberContainer;
      arrayContainer?: ArrayContainer;
    }   

    interface NumberContainer {           
      type: 'number';
      a: number;                                           
      b: number;       
    }     
      
    interface ArrayContainer {
      type: 'array';
      c: Array<any>;           
      d: Array<any>;
    }                                                     

    interface InferredContainerNoType<T> {
      y: {
        // This nested mapped type causes errors to not be caught by the compiler
        // (e.g. missing 'type' field)
        [key in keyof Pick<T, Exclude<keyof T, 'type'>>]: {
          value: T[key]
        };
      }
    }

    interface InferredANoType {
      x: {
        [key in keyof A]: InferredContainerNoType<Exclude<A[key], undefined>>;
      };
    }


    const inferredANoType_bad: InferredANoType = {
      x: {
        numberContainer: { // error Property 'a' is missing in type '{ b: { value: number; }; }' but required in type '{ a: { value: number; }; b: { value: number; }; }'.
          y: {
            b: { value: 2 },
          }
        },
        arrayContainer: { //error Property 'c' is missing in type '{ d: { value: number[]; }; }' but required in type '{ c: { value: any[]; }; d: { value: any[]; }; }'.
          y: {
            // Missing 'c' field
            d: { value: [2] },
          },
        },
      },
    };
```

**Comment by mikeecb** [comment]
@dragomirtitian thank you so much!
